src/main/java/org/gradle/profiler/studio/StudioGradleClient.java

```diff
@@ -58,34 +68,40 @@ public StudioGradleClient(GradleBuildConfiguration buildConfiguration, Invocatio
         commandLine.add("-Xbootclasspath/a:" + Joiner.on(File.pathSeparator).join(launchConfiguration.getSharedJars()));
         commandLine.add(launchConfiguration.getMainClass());
         commandLine.add(invocationSettings.getProjectDir().getAbsolutePath());
+        System.out.println("* Android Studio logs can be found at: " + Paths.get(launchConfiguration.getStudioPluginsDir().toString(), "idea.log"));
         System.out.println("* Using command line: " + commandLine);
 
+        new StudioPluginInstaller().installPlugin(launchConfiguration);
         return new CommandExec().inDir(studioInstallDir.toFile()).start(commandLine);
     }
 
     @Override
     public void close() throws IOException {
-        System.out.println("* PLEASE STOP ANDROID STUDIO....");
+        System.out.println("* Stopping Android Studio....");
+        studioPluginConnection.send(new StudioRequest(EXIT));
         studioProcess.waitForSuccess();
         System.out.println("* Android Studio stopped.");
-
-        server.close();
+        studioPluginServer.close();
+        studioAgentServer.close();
     }
 
-    public Duration sync(List<String> gradleArgs, List<String> jvmArgs) {
-        if (!hasRun) {
-            System.out.println("* PLEASE RUN SYNC IN ANDROID STUDIO (once it has finished starting up)....");
-            hasRun = true;
-        } else {
-            System.out.println("* PLEASE RUN SYNC IN ANDROID STUDIO....");
-        }
-
-        // Use a long time out because it can take quite some time between the tapi action completing and studio finishing the sync
-        SyncStarted started = agentConnection.receiveSyncStarted(Duration.ofMinutes(10));
-        agentConnection.send(new SyncParameters(gradleArgs, jvmArgs));
+    public BuildActionResult sync(List<String> gradleArgs, List<String> jvmArgs) {
+        System.out.println("* Running sync in Android Studio...");
+        studioPluginConnection.send(new StudioRequest(SYNC));
+        System.out.println("* Sent sync request");
+        // Use a long time out because it can take quite some time
+        // between the tapi action completing and studio finishing the sync
+        studioAgentConnection.receiveSyncStarted(Duration.ofMinutes(10));
```

Comment:
I think it would be good to extract all the wait times into constants in this class, to have an overview what we are waiting for and how long.

Reply:
done